Extension/src/LanguageServer/referencesProvider.ts

```diff
@@ -0,0 +1,97 @@
+        case ReferenceType.String: return util.getExtensionFilePath("assets/string.svg");
+        case ReferenceType.Inactive: return util.getExtensionFilePath("assets/cannotconfirm.svg");
+        case ReferenceType.CannotConfirm: return util.getExtensionFilePath("assets/cannotconfirm.svg");
+        case ReferenceType.NotAReference: return util.getExtensionFilePath("assets/not-a-reference.svg");
+    }
+    return util.getExtensionFilePath("assets/cannotconfirm.svg");
+}
+
+type TreeObject = FileItem | ReferenceItem | ReferenceTypeItem;
+
+export class ReferenceDataProvider implements vscode.TreeDataProvider<TreeObject> {
+    private references: Model;
+    private readonly _onDidChangeTreeData = new vscode.EventEmitter<TreeObject>();
+    readonly onDidChangeTreeData = this._onDidChangeTreeData.event;
+
+    constructor() {
+    }
+
+    setModel(results: ReferenceInfo[]): void {
+        this.references = new Model(results);
+        this._onDidChangeTreeData.fire();
+    }
+
+    clear(): void {
+        this.references = undefined;
+        this._onDidChangeTreeData.fire();
+    }
+
+    getTreeItem(element: TreeObject): vscode.TreeItem {
+        if (!this.references) {
+            return;
+        }
+
+        if (element instanceof ReferenceItem) {
+            const result: vscode.TreeItem = new vscode.TreeItem(element.text);
+            result.collapsibleState = vscode.TreeItemCollapsibleState.None;
+            result.iconPath = getReferneceTypeIconPath(element.type);
+            result.command = {
+                title: 'Open Reference',
+                command: 'C_Cpp.ShowReferencesItem',
+                arguments: [element]
+            };
+            return result;
+        }
+
+        if (element instanceof FileItem) {
+            const result: vscode.TreeItem = new vscode.TreeItem(element.uri);
+            result.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+            result.iconPath = vscode.ThemeIcon.File;
+            result.description = true;
+            return result;
+        }
+
+        if (element instanceof ReferenceTypeItem) {
+            const label: string = convertReferenceTypeToString(element.type, false, false);
+            const result: vscode.TreeItem = new vscode.TreeItem(label);
+            result.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+            result.iconPath = getReferneceTypeIconPath(element.type);
+            return result;
+        }
+    }
+
+    getChildren(element?: TreeObject | undefined): TreeObject[] {
+        if (!this.references) {
+            return;
+        }
+
+        if (element instanceof FileItem) {
+            return element.getReferences();
+        }
+
+        if (element instanceof ReferenceTypeItem) {
+            return element.getFiles();
+        }
+
+        return this.references.items;
+    }
+}
+
+// TODO: add provider for rename

```

Comment:
Would this be the right file to put 'rename' stuff in?  This is referencesProvider.ts.  Perhaps renaming the file or putting the provider in a different file would be better.

Reply:
Pending design of rename and separating of current code that's used for reference results and rename. Planning to check this in first for Colen and I to collaborate on.